algviz/parsers/structures.py

```diff
@@ -0,0 +1,207 @@
+    def __eq__(self, other):
+        # Usually we want to compare pointers by UID, not by the value of referent.
+        # However, these __eq__ methods are very useful for testing
+        return isinstance(other, Pointer) and other.referent == self.referent
+
+    def untablify(self, obj_table):
+        self.referent = obj_table[self.referent]
+
+class Array(collections.UserList, DataStructure):
+    """An array data structure"""
+    def __init__(self, *args, **kwargs):
+        collections.UserList.__init__(self, *args)
+        DataStructure.__init__(self, **kwargs)
+
+    def untablify(self, obj_table):
+        # UserList has its underlying list accessible as a member
+        self.data = [obj_table[elt] for elt in self.data]
+
+class _Singleton(abc.ABCMeta):
+    # Embarassingly copied from https://stackoverflow.com/questions/6760685
+    _instances = {}
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(_Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+class NullType(DataStructure, metaclass=_Singleton):
+    """A null value.  Similar to python's None.  Fun to make."""
+    def __init__(self):
+        super().__init__(uid="#null")
+        del self.metadata  # no static dictionary allowed
+
+    def untablify(self, obj_table):
+        pass
+
+    def __bool__(self):
+        return False
+
+    def __repr__(self):
+        return "Null"
+
+Null = NullType()
+
+class LinkedListNode(DataStructure):
+    def __init__(self, value, successor=Null):
+        self.value = value
+        self.successor = successor
+
+    def untablify(self, obj_table):
+        self.value = obj_table[self.value]
+        self.successor = obj_table[self.successor]
+
+class Graph(DataStructure):
+    def __init__(self, nodes, edges, **kwargs):
+        super().__init__(**kwargs)
+        self.nodes = nodes
+        self.edges = edges
+
+    def untablify(self, obj_table):
+        self.nodes = [obj_table[n] for n in self.nodes]
+        self.edges = [obj_table[e] for e in self.edges]
+
+class Node(DataStructure):
+    def __init__(self, data, **kwargs):
+        super().__init__(**kwargs)
+        self.data = data
+        # self.edges = edges  # no storing edges, I guess.  TODO: discuss this with people
+
+    def untablify(self, obj_table):
+        self.data = obj_table[self.data]
+        # self.edges = list(obj_table[e] for e in self.edges)
+
+class Edge(DataStructure):
+    def __init__(self, orig, dest, data=Null, **kwargs):
+        super().__init__(**kwargs)
+        self.orig = orig
+        self.dest = dest
+        self.data = data
+
+    def untablify(self, obj_table):
+        self.orig = obj_table[self.orig]
+        self.dest = obj_table[self.dest]
+        self.data = obj_table[self.data]
+
+class Widget(DataStructure):
+    # Potentially our most useful DataStructure
+
+    def __eq__(self, other):
+        return isinstance(other, Widget)
```

Comment:
why should all instances of widget be equal?

Reply:
I wanted two Arrays of Widgets to compare equal to each other in my test code.  But maybe I just shouldn't be using Widgets in my test cases when I compare Arrays for equality.

Reply:
I guess the real problem is that I didn't want __eq__ for anything but testing.  But that's silly, because equality testing is used for important stuff like hash tables.  Also, I thought it would be convenient for __eq__ to ignore "meta" information like UID and "metadata", e.g. so I could write test code for the particular test case:
```python
tree_1 = whatever_decode_method('''{"T": "tree", "root": {"T": "null"}}''')
# tree_1 has an auto-generated UID
tree_2 = structures.Tree(root=structures.Null)
# tree_2.uid is None
assert tree_1 == tree_2
```
Now, I think a test like that is actually just lazy.  I'd rather have to write:
```python
tree_2 = structures.Tree(root=structures.Null, uid=tree_1.uid)
assert tree_1 == tree_2
```
One easy consequence of this sort of equality testing is that I could immediately improve the implementation of graphs by putting the nodes and edges in a set and dict respectively.

Reply:
So, you are claiming that == should check that two objects have the same uid and represent the same structure? Is that particularly useful over DataStructure.same_object()?